File: code/backend/java/core/src/test/java/games/paths/core/model/match/MatchTraitCodecTest.java
package games.paths.core.model.match;

import org.junit.jupiter.api.Test;

import java.lang.reflect.Constructor;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Unit tests for {@link MatchTraitCodec}. Step 0.19.9 — exercises every
 * branch of the trait-uuid CSV join/split helpers.
 */
class MatchTraitCodecTest {

    @Test
    void joinNullOrEmptyReturnsNull() {
        assertNull(MatchTraitCodec.join(null));
        assertNull(MatchTraitCodec.join(new ArrayList<>()));
    }

    @Test
    void joinAllBlankReturnsNull() {
        assertNull(MatchTraitCodec.join(Arrays.asList(null, "", "   ")));
    }

    @Test
    void joinKeepsAndTrimsNonBlankValues() {
        assertEquals("a,b", MatchTraitCodec.join(Arrays.asList("a", null, "  b  ", "")));
    }

    @Test
    void splitNullOrBlankReturnsEmptyList() {
        assertTrue(MatchTraitCodec.split(null).isEmpty());
        assertTrue(MatchTraitCodec.split("   ").isEmpty());
    }

    @Test
    void splitParsesTrimsAndDropsBlanks() {
        assertEquals(List.of("a", "b"), MatchTraitCodec.split(" a , ,b, "));
    }

    @Test
    void splitResultIsMutable() {
        List<String> result = MatchTraitCodec.split("a");
        result.add("b");
        assertEquals(2, result.size());
    }

    @Test
    void privateConstructorForCoverage() throws Exception {
        Constructor<MatchTraitCodec> ctor = MatchTraitCodec.class.getDeclaredConstructor();
        ctor.setAccessible(true);
        assertNotNull(ctor.newInstance());
    }
}

File: code/backend/php/tests/Unit/Core/Domain/Matches/MatchTraitCodecTest.php
<?php

declare(strict_types=1);

namespace Games\Paths\Tests\Unit\Core\Domain\Matches;

use Games\Paths\Core\Domain\Matches\MatchTraitCodec;
use PHPUnit\Framework\TestCase;

class MatchTraitCodecTest extends TestCase
{
    public function testJoinNullOrEmptyReturnsNull(): void
    {
        $this->assertNull(MatchTraitCodec::join(null));
        $this->assertNull(MatchTraitCodec::join([]));
    }

    public function testJoinAllBlankReturnsNull(): void
    {
        $this->assertNull(MatchTraitCodec::join(['', '  ']));
    }

    public function testJoinKeepsAndTrimsNonBlankValues(): void
    {
        $this->assertSame('a,b', MatchTraitCodec::join(['a', '', '  b  ']));
    }

    public function testSplitNullOrBlankReturnsEmptyList(): void
    {
        $this->assertSame([], MatchTraitCodec::split(null));
        $this->assertSame([], MatchTraitCodec::split('   '));
    }

    public function testSplitParsesTrimsAndDropsBlanks(): void
    {
        $this->assertSame(['a', 'b'], MatchTraitCodec::split(' a , ,b, '));
    }
}
